File: src/prod/src/ServiceModel/ServiceCorrelationScheme.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace std;
using namespace Common;
using namespace ServiceModel;

void ServiceCorrelationScheme::WriteToTextWriter(__in TextWriter & w, ServiceCorrelationScheme::Enum const & val)
{
    switch (val)
    {
    case ServiceCorrelationScheme::Affinity:
        w << L"Affinity";
        return;
    case ServiceCorrelationScheme::AlignedAffinity:
        w << L"AlignedAffinity";
        return;
    case ServiceCorrelationScheme::NonAlignedAffinity:
        w << L"NonAlignedAffinity";
        return;
    default:
        Common::Assert::CodingError("Unknown ServiceCorrelationScheme value {0}", static_cast<int>(val));
    }
}

ServiceCorrelationScheme::Enum ServiceCorrelationScheme::GetScheme(std::wstring const & val)
{
    if (StringUtility::AreEqualCaseInsensitive(val, L"Affinity"))
    {
        return ServiceCorrelationScheme::Affinity;
    }
    else if (StringUtility::AreEqualCaseInsensitive(val, L"AlignedAffinity"))
    {
        return ServiceCorrelationScheme::AlignedAffinity;
    }
    else if (StringUtility::AreEqualCaseInsensitive(val, L"NonAlignedAffinity"))
    {
        return ServiceCorrelationScheme::NonAlignedAffinity;
    }
    else
    {
        Common::Assert::CodingError("Unknown ServiceCorrelationScheme value {0}", val);
    }
}

FABRIC_SERVICE_CORRELATION_SCHEME ServiceCorrelationScheme::ToPublicApi(__in ServiceCorrelationScheme::Enum const & scheme)
{
    switch (scheme)
    {
    case ServiceCorrelationScheme::Affinity:
        return FABRIC_SERVICE_CORRELATION_SCHEME_AFFINITY;
    case ServiceCorrelationScheme::AlignedAffinity:
        return FABRIC_SERVICE_CORRELATION_SCHEME_ALIGNED_AFFINITY;
    case ServiceCorrelationScheme::NonAlignedAffinity:
        return FABRIC_SERVICE_CORRELATION_SCHEME_NONALIGNED_AFFINITY;
    default:
        Common::Assert::CodingError("Unknown ServiceCorrelationScheme value {0}", static_cast<int>(scheme));
    }
}

ServiceCorrelationScheme::Enum ServiceCorrelationScheme::FromPublicApi(FABRIC_SERVICE_CORRELATION_SCHEME const & scheme)
{
    switch (scheme)
    {
    case FABRIC_SERVICE_CORRELATION_SCHEME_AFFINITY:
        return ServiceCorrelationScheme::Affinity;
    case FABRIC_SERVICE_CORRELATION_SCHEME_ALIGNED_AFFINITY:
        return ServiceCorrelationScheme::AlignedAffinity;
    case FABRIC_SERVICE_CORRELATION_SCHEME_NONALIGNED_AFFINITY:
        return ServiceCorrelationScheme::NonAlignedAffinity;
    default:
        Common::Assert::CodingError("Unknown ServiceCorrelationScheme value {0}", static_cast<int>(scheme));
    }
}

File: src/prod/src/managed/Api/src/System/Fabric/Description/ServiceCorrelationScheme.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Description
{
    /// <summary>
    /// <para>Indicates that this service is associated with another service, and describes the relationship with that service.</para>
    /// </summary>
    /// <remarks>
    ///   <para />
    /// </remarks>
    public enum ServiceCorrelationScheme
    {
        /// <summary>
        /// <para>An invalid correlation scheme. Cannot be used.</para>
        /// </summary>
        Invalid = 0,
        /// <summary>
        /// <para>Indicates that this service has an affinity relationship with another service. Provided for backwards compatibility, consider preferring 
        /// the Aligned or NonAlignedAffinity options.</para>
        /// </summary>
        Affinity = 1,
        /// <summary>
        /// <para>Aligned affinity ensures that the primaries of the partitions of the affinitized services are collocated on the same nodes. This is the 
        /// default and is the same as selecting the “Affinity” scheme.</para>
        /// </summary>
        AlignedAffinity = 2,
        /// <summary>
        /// <para>Non-Aligned affinity guarantees that all replicas of each service will be placed on the same nodes. Unlike Aligned Affinity, this does not 
        /// guarantee that replicas of particular role will be collocated. </para>
        /// </summary>
        NonAlignedAffinity = 3
    }
}